enhancements/openshift-console-cloud-shell.md

```diff
@@ -0,0 +1,195 @@
+- How does a Cloud Shell instance compares to an Eclipse Che workspace
+- Authentication and authorization mechanism to secure CloudShell instances
+- The mechanism to deny any user other then the owner to `exec` into a Cloud Shell Pod
+
+#### Cloud Shell vs Che Workspace
+
+A cloud shell instance will be implemented as a minimal Che workspace pod that includes the following 3 containers:
+
+- [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy): for authentication
+- Terminal server: connects to `xterm.js` and does `exec` to spawn shells in the tooling container 
+- Tooling container: with CLI tools like `bash`, `oc`, `odo`, `jq`, `vim` and stuff
+
+![](https://i.imgur.com/3mSAe1w.png)
+
+
+Such a workspace will be defined using a `DevWorkspace` Custom Resource (c.f. [#15425](https://github.com/eclipse/che/issues/15425)):
+
+```yaml
+apiVersion: workspaces.ecd.eclipse.org/v1alpha1
+kind: DevWorkspace
+metadata:
+  name: cloudshell
+spec:
+  components:
+    - type: dockerimage
+      id: eclipse/cloudshell/latest
+``` 
+
+As a consequence, the Cloud Shell will be dependent on the Eclipse Che DevWorkspace operator that deploys the CRD and controller to manage workspaces. But unlike Eclipse Che, in a Cloud Shell instance:
+- Eclipse Che server components are not deployed: `wsmaster`, `registries`, `keycloak` and `postgres`. Instanciation of `DevWorkspaces` is a responsibility of the OpenShift Console.
+- Eclipse Che and the Cloud Shell are both deployed and controlled using operators but those are distinct ones.
+- Theia (or other editors) based workspaces are not supported. Only Eclipse Che workspaces with editor of type CloudShell are supported.
+
+#### Authentication and Authorization
+
+The Terminal server endpoint is exposed through a TLS route. That's the endpoint that the clients running on the browser side (xterm.js) will connect to.
+
+![](https://i.imgur.com/XWcgHeI.png)
+
+The endpoint is secured through an [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy) sidecar. The client includes the user OpenShift token in its requests HTTP header. Only authenticated OpenShift users that satisfy the RBAC policy requirements are able to send requests to the Terminal server.
+
+In order to prevent unauthorized request to the Terminal server (coming from inside the OpenShift cluster) a [Network Policy](https://kubernetes.io/docs/concepts/services-networking/network-policies/) is used.
+
+#### Cloud Shell Pod Exec Denial Policy
+
+Users tokens are injected in the Tooling container. That's required for `oc` or `odo` to work properly. As a consequence access to the Tooling container should be allowed **only to the user of the Cloud Shell instace**. That's something that the authorization mechanism based on RBAC seen above is not able to garantee: users with editor privileges at cluster scope will be able to exec into the Pod.
```

Comment:
Can you describe the API interaction you will use to guarantee this?  Will it take advantage of certain aspects of pod immutability?  It cannot be an annotation.

Reply:
There is 
- A Mutating Admission Webhook that adds an annotation with the creator UID to the DevWorkspace CR at creation time;

Then WorkspaceController creates Deployment with that creator annotation in `metadata.annotations` + `spec.template.metadata.annotations`. As a result, deployment creates pods with creator annotation propagated.

- A Mutating Admission Webhook that validates that the creator annotation is not modified in workspace-related deployments and pods. If the value is omit - it's restored from the previous object state, if it's modified - error is thrown. 

- A Validating Admission Webhook that validates exec to make sure that the user who request connect to the workspace-related pod is the same as pod creator annotation.

Hope it answers: Can you describe the API interaction you will use to guarantee this? 
If not - let me know.

> Will it take advantage of certain aspects of pod immutability?

Do you mean that it would be better to propagate creator info like through Env vars, or command/args, or other fields which is immutable?
We haven't thought about it before. Will take into account.